### PR TITLE
Dixon

### PR DESCRIPTION
**Pull request by jgdumas**
[+] Change examples/solverrat.C into examples/dixondenseLU to reflect Dense System Solving using LU decomposition over a prime field and Dixon's lifting over the rationals.
[+] Start examples/dixonsparseelim.C 
[+] clean-up sparsesolverat/densesolverat as Direct solvers

Example: for SIMC BasisLib/problems/qap10.sms on one core of a i7-6700 CPU @ 3.40GHz
> ./dixondenseLU qap10.sms
CPU time (seconds): 0.989956
> ./dixonsparseelim qap10.sms
CPU time (seconds): 0.669377
> ./densesolverat qap10.sms
CPU time (seconds): 5.61988
> ./sparsesolverat qap10.sms
CPU time (seconds): 41.3938

### CODE REVIEW

**Comment by romainlebreton** [comment]
Need review

Only TODO remaining : Maple format for sparseMatrix could have linebreaks after each row for nicer better printring

File: examples/dixonsolve.C [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+#include "linbox/matrix/dense-matrix.h"
+
+// SparseElim
+#include "linbox/matrix/sparse-matrix.h"
+
+using namespace LinBox;
+typedef Givaro::ZRing<Givaro::Integer> Ints;
+typedef DenseVector<Ints> ZVector;
+
+
```
Comment (ClementPernet) [inline examples/dixonsolve.C:48]: This constant prime iterator should be added in  `randiter/random-prime.h` which already has some specialization of prime iterators such as the prime sequence adaptor.
Let's avoid to write utility functions in example files, especially when it is shared by 2 files!

File: examples/dixonsolve.C [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+        //====================================================
+        // BEGIN Replacement solve with fixed prime
+
+    Ints::Element d;
+    Method::Dixon m(M);
+    typedef Givaro::Modular<double> Field;
+		// 0.7213475205 is an upper approximation of 1/(2log(2))
+    size_t bitsize((size_t)( 26-(int)ceil(log((double)A.rowdim())*0.7213475205)));
```
Comment (ClementPernet) [inline examples/dixonsolve.C:146]: Use `FieldTraits<Givaro::Modular<double>>::betBitSize(A.rowdim())` instead (defined in `field/field-traits.h`

File: examples/dixonsparseelim.C [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        // BlasElimination
+    Method::SparseElimination M;
+    M.singularity = Singularity::NonSingular;
+
+        //====================================================
+        // BEGIN Replacement solve with fixed prime
+    Method::Dixon m(M);
+    typedef Givaro::Modular<double> Field;
+		// 0.7213475205 is an upper approximation of 1/(2log(2))
+    size_t bitsize((size_t)( 26-(int)ceil(log((double)A.rowdim())*0.7213475205)));
```
Comment (ClementPernet) [inline examples/dixonsparseelim.C:140]: use `FieldTraits<Givaro::Modular<double>>::betBitSize(A.rowdim())` instead (defined in `field/field-traits.h`)

File: examples/dixonsparseelim.C [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+#include "linbox/util/matrix-stream.h"
+#include "linbox/solutions/methods.h"
+#include "linbox/util/error.h"
+
+using namespace LinBox;
+typedef Givaro::ZRing<Givaro::Integer> Ints;
+typedef DenseVector<Ints> ZVector;
+
+
+struct FixPrime {
```
Comment (ClementPernet) [inline examples/dixonsparseelim.C:41]: same here: factor out this code and adapt it to the PrimeIterator API  in random-prime.h. for instance creating a new `IteratorCategoris::ConstantTag`

File: examples/dixonsparseelim.C [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    ZVector X(ZZ, A.coldim()),B(ZZ, A.rowdim());
+    ZVector U(ZZ, A.coldim());
+    
+    // B = A U
+    if (createB) {
+        std::cerr << "Creating a random {-1,1} vector " << std::endl;
+        srand48( BaseTimer::seed() );
+        for(ZVector::iterator it=B.begin();
+            it != B.end(); ++it)
+            if (drand48() <0.5)
```
Comment (ClementPernet) [inline examples/dixonsparseelim.C:107]: I thought our clarification of the randiter interface would have made possible to write somthing like 
```
ZRing<Integer>::RandIter G(2)
B.random(G);
``` 
but it seems that we do not offer uniform sampling  in intervals not starting in 0 in Givaro. Too bad.

File: examples/dixonsolve.C (hunk 1 above)
Comment (jgdumas) [reply examples/dixonsolve.C:50]: ok. done.

File: examples/dixonsolve.C (hunk 2 above)
Comment (jgdumas) [reply examples/dixonsolve.C]: ok done.

File: examples/dixonsparseelim.C (hunk 3 above)
Comment (jgdumas) [reply examples/dixonsparseelim.C]: dixonsparselim removed (now both dense and sparse in dixonsolve)

File: examples/dixonsparseelim.C (hunk 4 above)
Comment (jgdumas) [reply examples/dixonsparseelim.C]: idem.

File: examples/dixonsparseelim.C (hunk 5 above)
Comment (jgdumas) [reply examples/dixonsparseelim.C]: Now (with bool instead of Integer) in PR  linbox-team/givaro#196

**Comment by ClementPernet** [comment]
Ok, good to go.

**Pull request merged**